t/t3907-stash-show-config.sh

```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
+
+test_description='Test git stash show configuration.
+
+'
+
+. ./test-lib.sh
+
+config_stash_stat_and_patch () {
+	if test "$1" != "unset"
+	then
+		git config stash.showStat "$1"
+	fi
+
+	if test "$2" != "unset"
+	then
+		git config stash.showPatch "$2"
+	fi
+}
+
+unset_config () {
+	if git config --bool stash.showStat >/dev/null
+	then
+		git config --unset stash.showStat
+	fi
+
+	if git config --bool stash.showPatch >/dev/null
+	then
+		git config --unset stash.showPatch
+	fi
+}
+
+get_expected_output () {
+	FLAGS=
+
+	if test "$(git config --bool stash.showStat || echo true)" = "true"
+	then
+		FLAGS=--stat
+	fi
+
+	if test "$(git config --bool stash.showPatch || echo false)" = "true"
+	then
+		FLAGS=${FLAGS}${FLAGS:+ }-p
+	fi
+
+	git diff $FLAGS >expect
+}
+
+get_actual_output () {
+	git stash &&
+	git stash show >actual
+}
+
+check_output () {
+	get_expected_output &&
+	get_actual_output &&
+	test_cmp expect actual &&
+	rm expect actual
+}
+
+options="unset true false"
+
+test_expect_success 'setup' '
+	echo 1 >file &&
+	git add file &&
+	git commit -m initial
+'
+
+for i in $options
+do
+	for j in $options
+	do
+		if test "$i" = "false" && (test "$j" = "false" || test "$j" = "unset")
+		then
+			test_expect_success "showStat $i and showPatch $j" '
+				config_stash_stat_and_patch $i $j &&
+				echo 2 >file &&
+				get_actual_output &&
+				test_must_be_empty actual &&
+				rm actual
+			'
+			continue
+		fi
+
+		unset_config
+		test_expect_success "showStat $i and showPatch $j" '
+			config_stash_stat_and_patch $i $j &&
+			echo 2 >file &&
+			check_output
+		'
+	done
+done
```

Comment:
Okay, so I think I would prefer something like this instead:

```sh
#!/bin/sh

test_description='Test git stash show configuration.'

. ./test-lib.sh

test_expect_success 'setup' '
	test_commit file
'

# takes three parameters:
# 1. the stash.showStat value (or "<unset>")
# 2. the stash.showPatch value (or "<unset>")
# 3. the diff options of the expected output (or "<empty>" for no output)
test_stat_and_patch () {
	if test "<unset>" = "$1"
	then
		git config --unset stash.showStat
	else
		test_config stash.showStat "$1"
	fi &&
	if test "<unset>" = "$2"
	then
		git config --unset stash.showPatch
	else
		test_config stash.showPatch "$2"
	fi &&

	echo 2 >file.t &&
	git diff "$@" >expect &&
	git stash &&
	git stash show >actual &&

	shift &&
	shift &&
	if test "<empty>" = "$1"
	then
		test_must_be_empty actual
	else
		test_cmp expect actual
	fi
}

test_expect_success 'various combinations' '
	test_stat_and_patch "<unset>" "<unset>" &&
	test_stat_and_patch "<unset>" false &&
	test_stat_and_patch "<unset>" true -p &&
	test_stat_and_patch false "<unset>" "<empty>" &&
	test_stat_and_patch false false "<empty>" &&
	test_stat_and_patch false true -p &&
	test_stat_and_patch true "<unset>" --stat &&
	test_stat_and_patch true false --stat &&
	test_stat_and_patch true true --stat -p
'

test_done
```

I could even imagine that that `<empty>` handling is totally unnecessary...

What do you think?

Reply:
Thanks, @dscho! Indeed, your approach makes debugging easier. To make it even easier to debug, can't we do a `test_stat_and_patch` per `test_expect_success`?. 

The first two `if`s inside `test_patch_success` should be "unchained" since `git config --unset foobar` returns non-zero value if `foobar` is not set in the first place. I guess that, in this case, a break in `&&`-chain wouldn't be a problem, right? I mean, if `git config` doesn't work as intended, then it would be reported by `t1300-config.sh`

I made some more little changes to it (call one `test_stat_and_patch` per `test_expect_sucesss`, moved `shift && shift` before `git diff $@` and dropped `<empty>` handling since it is not necessary)  and force-pushed it. Do you think it is good enough now?

Reply:
> The first two ifs inside test_patch_success should be "unchained" 

Does that make it past chainlint?  I guess it might because this is being done in a function?  You could also consider using [`test_might_fail`](https://github.com/git/git/blob/master/t/README#L717), which is happy with both successes and failures, but will catch segfaults.